internal/cache/filesystem.go

```diff
@@ -1,53 +1,92 @@
 package cache
 
 import (
+	"context"
 	"fmt"
 	"io/ioutil"
 	"os"
 	"path"
 	"time"
+
+	"github.com/pkg/errors"
+)
+
+const (
+	cacheSubdir = "parrot"
 )
 
-type FileCache struct {
+type FilesystemCache struct {
 	dir string
 	ttl time.Duration
 }
 
-func NewFileCache(ttl time.Duration) (*FileCache, error) {
+func NewFilesystemCache(ttl time.Duration) (*FilesystemCache, error) {
 	dir, err := os.UserCacheDir()
 	if err != nil {
-		return nil, err
+		return nil, errors.Wrap(err, "Failed to determin OS user cache directory")
 	}
-	cacheDir := path.Join(dir, "parrot")
+
+	cacheDir := path.Join(dir, cacheSubdir)
+
 	err = os.MkdirAll(cacheDir, os.ModeDir)
 	if err != nil {
-		return nil, err
+		return nil, errors.Wrap(err, "Failed to create cacge directory")
 	}
-	return &FileCache{
+
+	return &FilesystemCache{
 		dir: cacheDir,
 		ttl: ttl,
 	}, nil
 }
 
-func (f *FileCache) GetTranslation(projectID int, languageCode, format string) ([]byte, error) {
-	s, err := os.Stat(f.key(projectID, languageCode, format))
-	if os.IsNotExist(err) {
-		return nil, ErrCacheMiss
+func (f *FilesystemCache) GetTranslation(ctx context.Context, projectID int, languageCode, format string) ([]byte, error) {
+	filename := f.key(projectID, languageCode, format)
+
+	s, err := os.Stat(filename)
+	if err != nil {
+		if os.IsNotExist(err) {
+			return nil, ErrCacheMiss
+		}
+
+		return nil, errors.Wrap(err, "Failed to get cached file state from OS")
```

Comment:
```suggestion
	if os.IsNotExist(err) {
		return nil, ErrCacheMiss
	} else if err != nil {
		return nil, errors.Wrap(err, "Failed to get cached file state from OS")
```

Reply:
Seems weird to me asking "Is THIS of type x" and then ask "does THIS even exists"

Reply:
I don't think of it like that. You've already assigned `err` a value. nil just happens to be `no error`. I think the nested ifs make it messier to read.

Reply:
Okay, in that case maybe we should just remove "else if" and just have "if" ? this should also cleanup the control flow a tiny bit.

Reply:
@naueramant
![](https://64.media.tumblr.com/tumblr_marjdqJISc1qgzqnco10_r1_250.gif)